src/client/client_impl.cc

```diff
@@ -288,21 +320,256 @@ google::protobuf::util::StatusOr<std::pair<size_t, void*>> ClientImpl::ReadFile(
+          "Send data encountered irrecoverable failure");
+    }
+
+    if (send_data_recoverable_error.load()) {
+      retry--;
+      continue;
+    }
+
+    // Prepare WriteFileChunkRequest, WriteFileChunkRequestHeader
+    WriteFileChunkRequest write_request;
+    write_request.mutable_header()->set_chunk_handle(chunk_handle);
+    write_request.mutable_header()->set_chunk_version(chunk_version);
+    write_request.mutable_header()->set_offset_start(offset);
+    write_request.mutable_header()->set_length(nbytes);
+    write_request.mutable_header()->set_data_checksum(data_checksum);
+    for (auto location : chunk_server_location_entry.locations) {
+      write_request.mutable_replica_locations()->Add(std::move(location));
+    }
+    // Prepare the client context    
+    grpc::ClientContext client_context;
+    common::SetClientContextDeadline(client_context, config_manager_);
+ 
+    // We send this WriteFileChunkRequest to the primary 
+    auto primary_location(chunk_server_location_entry.primary_location);
+    const std::string primary_server_address(
+        primary_location.server_hostname() + ":" + 
+            std::to_string(primary_location.server_port()));
+    
+    auto primary_server_service_client(GetChunkServerServiceClient(
+                                           primary_server_address));
+    // Issue WriteFileChunkRequest and check status
+    StatusOr<WriteFileChunkReply> write_reply_or(
+        primary_server_service_client->SendRequest(write_request, client_context));
+
+    // Handle grpc error, and retry if possible
+    if (!write_reply_or.ok()) {
+      LOG(ERROR) << "Send write chunk data request to " 
+                 << primary_server_address << "failed due to " 
+                 << write_reply_or.status();
+    } else {
+      auto write_reply(write_reply_or.ValueOrDie());
+      switch (write_reply.status()) {
+        case FileChunkMutationStatus::OK:
+          LOG(INFO) << "Write to file " << filename << " at chunk_index "
+                    << chunk_index << " and offset " << offset << " for "
+                    << nbytes << " bytes succeeds";
+          return write_reply_or;
+        case FileChunkMutationStatus::FAILED_NOT_LEASE_HOLDER:
+          // Client found out that the primary it though is no longer a lease
+          // holder, we need to force to get the chunk metadata and retry
+          LOG(INFO) << "Retrying as primary server "
+                    << primary_server_address << " is no longer lease holder "
+                    << "refetching chunk metadata";
+          get_metadata_status = 
+              GetMetadataForChunk(filename, chunk_index, 
+                                  OpenFileRequest::WRITE, chunk_handle, 
+                                  chunk_version,
+                                  chunk_server_location_entry, true);
+          if (!get_metadata_status.ok()) {
+            return get_metadata_status; 
```

Comment:
should we log the error  here for debug?

Reply:
Yes, definitely.